data_analysis/near-deduplication/minhash_deduplication_alt.py

```diff
@@ -5,697 +5,331 @@
 
-lsh: MinHashLSH | None = None
-dup_ids: Set[int] | None = None
-
-
-def load_dataset_with_config(conf: Dict[str, Any]) -> Dataset:
+def ngrams(sequence: List[str], n: int) -> Iterable:
     """
-    Load a dataset based on the configuration. Be careful about changing this function,
-    as it is used for caching the intermediate results.
+    Directly taken from nltk package to avoid dependency.
 
     Parameters
     ----------
-    conf : Dict[str, Any]
-        The configuration. Mainly, there are three ways to load a dataset:
-        1. Directly from th ehub
-        2. From a local git repository
-        3. From a local dataset directory that was saved by `save_to_disk` before
+    sequence : list
+        The sequence of items to be n-grammed.
+    n : int
+        The order of the n-grams to be extracted.
 
     Returns
     -------
-    Dataset
-        The loaded dataset.
+    Iterable
+        The n-grams generated from the sequence.
     """
+    iterables = tee(sequence, n)
+    for i, sub_iterable in enumerate(iterables):
+        for _ in range(i):
+            next(sub_iterable, None)
+    return zip(*iterables)
 
-    # Load from hub
-    if not conf["lfs"]:
-        ds = load_dataset(
-            conf["dataset"],
-            conf["config"],
-            data_dir=conf["data_dir"],
-            split=conf["split"],
-            use_auth_token=True,
-            cache_dir=conf["cache_dir"],
-        )
-    # Or load from git lfs files, if there isn't a local concatenated dataset
-    elif not os.path.exists(conf["concat_output"]):
-        datasets = []
-        # In practice, it might stuck here, you can hit Ctrl+C and run it again.
-        for file in tqdm(sorted(glob.glob(conf["data_dir"] + "/*.jsonl")), desc="Loading datasets..."):
-            datasets.append(load_dataset("json", data_files=file, split=conf["split"], cache_dir=conf["cache_dir"]))
-        ds = concatenate_datasets(datasets)
-        ds.save_to_disk(conf["concat_output"])
-        ds = load_from_disk(conf["concat_output"])
-    # Or load from the concatenated dataset
-    else:
-        ds = load_from_disk(conf["concat_output"])
-
-    # Assign unique index to each record
-    # A token length filtering was used in the Python experiment
-    ds = ds.filter(
-        lambda x: len({t for t in NON_ALPHA.split(x[conf["column"]]) if t}) >= conf["min_token_length"],
-        num_proc=os.cpu_count(),
-        desc="Filtering records...",
-    )
-    ds = ds.map(
-        lambda _, idx: {"__id__": idx},
-        with_indices=True,
-        num_proc=os.cpu_count(),
-        desc="Adding index...",
-    )
-
-    return ds
-
-
-def embed_func(idx: int, content: str, *, num_perm: int) -> Dict[str, Any]:
+
+def sha1_hash32(data):
     """
-    Embed the content of a record into a MinHash object. This function should be
-    used with multiprocessing and it scales well with the number of cores.
+    Directly taken from datasketch package to avoid dependency.
```

Comment:
Is it really necessary? Having a dependency is not a big deal imho

Reply:
Having a dependency here isn't an issue. The main issue is that the high-level abstraction offered by that package isn't flexible enough for parallelization.